### PR TITLE
ACSetSpec PEG.jl-based Implemenation

### PR DESCRIPTION
**Pull request by cscaff**
The following PR contains two main files: src/parsers.jl as well as test/parsers.jl. This contains a module for constructing a parse expression grammar for creating an ADT representation from an ACSet specification. The use of PEG.jl allows us to build ASTs that are not constrained to the way Julia treats code as data. We can define our expressions in our tree the way we see fit. I have also included a testing module to ensure that the parser functions properly. The new parser can be called by a string macro: "acsetspec_str". For instance, we can call:
```julia
acsetspec"""
    LabeledGraph{Symbol}
    begin
      V(label=a)
      V(label=b)
      V(label=c)
      E(src=1,tgt=3)
      E(src=2,tgt=3)
    end"""
 ```

### CODE REVIEW

**Comment by lukem12345** [comment]
I noticed that the function `parse_identifier` happened to always be passed `Chars`, so I devised a test for the case in which an identifier is more than one character.

This test fails:
```julia
@test args("foo, b ,c")[1] == ([Value(:foo), Value(:b), Value(:c)])
```
The LHS evaluates to `ACSets.ADTs.Value{Symbol}[f, b, c]`.

File: src/parsers.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+# Ensures "head" exists but does not check type
+@rule head = r"\S*"p |> v -> Symbol(v)
+# Block contains one or more lines of statements
+@rule block = line[*] & r"\n?"p |> v -> v[1]
+# Line contains a statement followed by a new line or ";"
+@rule line  = ws & statement & r"[^\S\r\n]*" & EOL |> v -> v[2]
+# Statement contains a call followed by arguments in parenthesis: "identifier(args)"
+@rule statement = identifier & lparen & args & rparen |> v -> Statement(Symbol(v[1]), v[3])
+# args contains one or more arguments separated by commas
+@rule args = (arg & ws & comma)[*] & arg |> v -> collect_args(v)
```
Comment (lukem12345) [inline src/parsers.jl:69]: We should make sure that whitespace is allowed before identifiers as in `( foo , bar , buz)`.

File: src/parsers.jl
Comment (lukem12345) [inline src/parsers.jl:1]: Use two-spaces instead of four-spaces for indentation. You can configure your IDE to do this for Julia files by default.

Comment (lukem12345) [inline src/parsers.jl:1]: Use `git mv` to capitalize this filename to match the style of this repo. Likewise for the corresponding test file.

File: src/parsers.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    output = Vector{Args}(first.(v[1]))
+    push!(output, last(v))
+end
+
+# Parses an identifier into a symbol/integer
+parse_identifier(v) = begin 
+   if isnothing(tryparse(Int, string(v)))
+        return Value(Symbol(v))
+   else
+        return Value(tryparse(Int, string(v)))
```
Comment (lukem12345) [inline src/parsers.jl:86]: There is a double-parse happening here. Use a local variable for `tryparse(Int, string(v))`.

File: src/parsers.jl (hunk 1 above)
Comment (cscaff) [reply src/parsers.jl]: Thanks for the feedback! I have addressed all the issues you mentioned. Looks all good! Let me know if there is anything else as well.

**Review by quffaro — APPROVED** [review]

**Comment by lukem12345** [comment]
@jpfairbanks It looks like @quffaro and I both approve. Merge?

**Comment by jpfairbanks** [comment]
Merge

**Pull request merged**